### PR TITLE
Enable SPI functionality in GraalVM native mode

### PR DESCRIPTION
**Pull request by pivovarit**
This change introduces a GraalVM substitution for `ServiceLoader#getServiceDefinitions` so that discovered classpath resources are properly discovered, read and not accidentally discarded.

### Motivation

Hazelcast SPI relies on the classpath presence of multiple _META-INF/services/com.foo.Bar_ files (for example, _com.hazelcast.spi.discovery.multicast.DiscoveryStrategyFactory_).

The internal implementation of _ServiceLoader_ processes and collects URLs to the classpath resources:

https://github.com/hazelcast/hazelcast/blob/d54fa6df9a88c59f7479744c2b46158f10b2df7f/hazelcast/src/main/java/com/hazelcast/internal/util/ServiceLoader.java#L101-L124

However, this implementation doesn't play along with GraalVM's native images since Substrate doesn't really feature a _normal_ classpath nor classloading but a mere _simulation_ of these.

URLs to classpath resources located in different jars normally look like:
```
jar:file:/work/application/lib/com.hazelcast.hazelcast-4.0.1.jar!/META-INF/services/com.hazelcast.spi.discovery.DiscoveryStrategyFactory
jar:file:/work/application/lib/com.hazelcast.hazelcast-kubernetes-2.0.1.jar!/META-INF/services/com.hazelcast.spi.discovery.DiscoveryStrategyFactory
```

but under Substrate, they become _resources_ and share the same URL (but `URL#openStream` leads to different content) which requires special handling:
```
resource:META-INF/services/com.hazelcast.spi.discovery.DiscoveryStrategyFactory
resource:META-INF/services/com.hazelcast.spi.discovery.DiscoveryStrategyFactory
```

So, not only extra processing and conversions are not welcome:
https://github.com/hazelcast/hazelcast/blob/d54fa6df9a88c59f7479744c2b46158f10b2df7f/hazelcast/src/main/java/com/hazelcast/internal/util/ServiceLoader.java#L109-L112

But also storing URLs in a set causes rejection of valid URLs:
https://github.com/hazelcast/hazelcast/blob/d54fa6df9a88c59f7479744c2b46158f10b2df7f/hazelcast/src/main/java/com/hazelcast/internal/util/ServiceLoader.java#L106

----

fixes https://github.com/hazelcast/quarkus-hazelcast-client/issues/18

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -78,3 +78,4 @@ To make testing simple, the extension provides the `HazelcastServerTestResource`
 ## Limitations (native mode)
 - Default Java serialization is not supported
 - User code deployment is not supported
+- Hazelcast SPI support is limited
```
Comment (leszko) [inline README.md:81]: Could you mention how it's limited? Does Kubernetes work? Do custom SPIs work?

File: runtime/src/main/java/io/quarkus/hazelcast/client/runtime/graal/ServiceLoaderUtils.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.hazelcast.internal.nio.IOUtil.closeResource;
+
+final class ServiceLoaderUtils {
+    private ServiceLoaderUtils() {
+    }
+
+    static Set<Target_ServiceDefinition> parse(URL url, ClassLoader classLoader, ILogger logger) {
```
Comment (leszko) [inline runtime/src/main/java/io/quarkus/hazelcast/client/runtime/graal/ServiceLoaderUtils.java:19]: 1. Could you add javadoc explaining what this method does? Or maybe make this method name more explanatory?
2. Wouldn't it be worth to add some unit tests for his class and other you added?

File: README.md (hunk 1 above)
Comment (pivovarit) [reply README.md:81]: To be honest, it's hard to define. The code from `parse` method is effectively a normal IMDG code stripped from some _magic_ needed for the sake of OSGi and Glassfish compatibility so I expect some problems in these areas provided that you can make GraalVM work with that.

Naturally, all standard SPI-enabled extensions provided at build time do work properly

File: runtime/src/main/java/io/quarkus/hazelcast/client/runtime/graal/ServiceLoaderUtils.java (hunk 2 above)
Comment (pivovarit) [reply runtime/src/main/java/io/quarkus/hazelcast/client/runtime/graal/ServiceLoaderUtils.java:19]: I will add docs around the Graal substitution and explain the problem there as well.

I'm working on tests at the moment. Unit testing is useless, we need an integration test that loads a custom extension via SPI

Comment (pivovarit) [reply runtime/src/main/java/io/quarkus/hazelcast/client/runtime/graal/ServiceLoaderUtils.java:23]: Actually, this is covered by Hasan's test suite

**Pull request merged**